lib/descriptor/common.js

```diff
@@ -0,0 +1,259 @@
+const common = exports;
+
+/**
+ * parse script context for descriptor
+ * @const {Object}
+ */
+
+common.scriptContext = {
+  TOP: 'TOP',
+  P2SH: 'P2SH',
+  P2WPKH: 'P2WPKH',
+  P2WSH: 'P2WSH',
+  P2TR: 'P2TR'
+};
+
+/**
+ * Derive type for a key object in a descriptor
+ * @const {Object}
+ */
+
+common.deriveType = {
+  NO: 'NO',
+  UNHARDENED: 'UNHARDENED',
+  HARDENED: 'HARDENED'
+};
+
+/**
+ * String type for key object in a descriptor
+ * @const {Object}
+ */
+
+common.stringType = {
+  PUBLIC: 'PUBLIC',
+  PRIVATE: 'PRIVATE',
+  NORMALIZED: 'NORMALIZED'
+};
+
+/**
+ * Output types for descriptor
+ * @const {Object}
+ */
+
+common.outputTypes = types;
+
+/**
+ * Valid input characters for input in a descriptor
+ * and used for calculating checksum
+ * @const {String}
+ */
+
+common.INPUT_CHARSET = '0123456789()[],\'/*abcdefgh@:$%{}IJKLMNOPQRSTUVWXYZ' +
+                       '&+-.;<=>?!^_|~ijklmnopqrstuvwxyzABCDEFGH`#"\\ ';
+
+ /**
+ * Checksum character set for calculating checksum
+ */
+
+common.CHECKSUM_CHARSET = 'qpzry9x8gf2tvdw0s3jn54khce6mua7l';
+
+/**
+ * Test whether this descriptor is of scriptType
+ * ('pk', 'pkh', 'wpkh' etc.) or not
+ * @param {String} scriptType type of script
+ * @param {String} desc descriptor string
+ * @returns {Boolean}
+ */
+
+common.descType = function descType(scriptType, desc) {
+  if (
+    desc.length >= scriptType.length + 2 &&
+    desc[scriptType.length] === '(' &&
+    desc[desc.length - 1] === ')' &&
+    scriptType === desc.substring(0, scriptType.length)
+  ) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Strip outer level script string from descriptor
+ * @param {String} str
+ * @param {String} s
+ * @returns {String} stripped descriptor string
+ */
+
+common.strip = function strip(scriptType, desc) {
+  desc = desc.substring(scriptType.length + 1, desc.length - 1);
+  return desc;
+};
+
+/**
+ * Extract the expression that str begins with.
+ * This function will return the initial part of str, up to
+ * (but not including) the first comma or closing brace,
+ * skipping ones that are surrounded by braces. So for example,
+ * for "hello(world)(),(" the initial part
+ * "hello(world)()" will be returned. sp will be
+ * updated to skip the initial part that is returned.
+ * @param {String} str
+ * @returns {String}
+ */
+
+common.giveExpr = function giveExpr(str) {
+  let level = 0;
+  let it = 0;
+
+  while (it < str.length) {
+    if (str[it] === '(' || str[it] === '{') {
+      level++;
+    } else if (level && (str[it] === ')' || str[it] === '}')) {
+      level--;
+    } else if (
+      level === 0 &&
+      (str[it] === ')' || str[it] === '}' || str[it] === ',')
+    ) {
+      break;
+    }
+    it++;
+  }
+
+  const ret = str.slice(0, it);
+  return ret;
+};
+
+/**
+ * Test whether a string is a hex string or not
+ * @param {String} str
+ * @returns {Boolean}
+ */
+
+common.isHex = function isHex(str) {
+  const regexp = /^[0-9a-fA-F]+$/;
+  return regexp.test(str);
+};
+
```

Comment:
Maybe add a reference to Core.
```suggestion

/**
  * internal function that computes the descriptor checksum
  * Ref https://github.com/bitcoin/bitcoin/blob/5e1aab23347eb02fa316cde82531a57327098ca7/src/bech32.cpp#L127-L216
  */
  
```